File: python/examples/string-pad_test.py
"""string-pad: str.center / ljust / rjust の Contract を検証する。"""

import pytest


def test_center_pads_both_sides() -> None:
    """左右に埋めて指定幅にする。既定の埋め文字は半角スペース。"""
    assert "abc".center(8) == "  abc   "
    assert "abc".center(8, "*") == "**abc***"
    assert "abc".center(7, "*") == "**abc**"


def test_input_is_not_mutated() -> None:
    """元の文字列は変わらない（新しい文字列を返す）。"""
    s = "abc"
    result = s.center(8)
    assert s == "abc"
    assert result is not s


def test_odd_remainder_side_depends_on_width_parity() -> None:
    """余りが奇数のとき、幅が偶数なら右、奇数なら左に 1 文字多く付く。"""
    assert "abc".center(4) == "abc "
    assert "a".center(4) == " a  "
    assert "ab".center(5) == "  ab "
    assert "abcd".center(7) == "  abcd "


def test_fstring_center_always_leans_right() -> None:
    """f-string の ^ 指定は余りを常に右に寄せるので、幅が奇数のとき str.center と食い違う。"""
    assert f"{'ab':^5}" == " ab  "
    assert "ab".center(5) == "  ab "
    assert f"{'abc':^4}" == "abc "
    assert "abc".center(4) == "abc "


def test_width_not_greater_than_length_returns_original() -> None:
    """幅が長さ以下（0・負数を含む）なら元の文字列をそのまま返し、切り詰めない。"""
    assert "abc".center(3) == "abc"
    assert "abc".center(2) == "abc"
    assert "abc".center(0) == "abc"
    assert "abc".center(-1) == "abc"


def test_empty_string_is_filled_entirely() -> None:
    """空文字は埋め文字だけで幅を満たす。"""
    assert "".center(4, "*") == "****"
    assert "".center(4) == "    "


def test_fillchar_must_be_exactly_one_char() -> None:
    """埋め文字は 1 文字ちょうどでなければ TypeError。"""
    with pytest.raises(TypeError):
        "abc".center(8, "ab")
    with pytest.raises(TypeError):
        "abc".center(8, "")


def test_width_must_support_index_protocol() -> None:
    """幅は __index__ を持つ値（int など）。float や str は TypeError。"""

    class Width:
        def __index__(self) -> int:
            return 8

    assert "abc".center(Width()) == "  abc   "
    with pytest.raises(TypeError):
        "abc".center(8.0)  # type: ignore[arg-type]
    with pytest.raises(TypeError):
        "abc".center("8")  # type: ignore[arg-type]


def test_length_is_counted_in_code_points() -> None:
    """長さはコードポイントで数える。絵文字・全角は 1、結合文字は 2。"""
    assert "🐶".center(4, "*") == "*🐶**"
    assert "日本".center(6, "*") == "**日本**"
    assert "é".center(4, "*") == "*é*"


def test_ljust_and_rjust_pad_one_side() -> None:
    """ljust は右に、rjust は左にだけ埋める。"""
    assert "abc".ljust(6, "*") == "abc***"
    assert "abc".rjust(6, "*") == "***abc"
    assert "abc".ljust(2) == "abc"
    assert "abc".rjust(2) == "abc"

File: ruby/examples/string-pad_test.rb
# カード string-pad の Contract を検証するテスト（String#center）
require "minitest/autorun"

class StringPadTest < Minitest::Test
  def test_pads_both_sides
    # 左右に padstr を詰めて指定幅にする
    assert_equal "  abc   ", "abc".center(8)
    assert_equal "**abc***", "abc".center(8, "*")
    assert_equal "abc***", "abc".ljust(6, "*")
    assert_equal "***abc", "abc".rjust(6, "*")
  end

  def test_returns_new_string_without_mutating
    # 元の文字列は変わらず、幅以下でも同じオブジェクトは返さない
    s = "abc"
    result = s.center(8)
    assert_equal "abc", s
    assert_equal "abc", s.center(3)
    refute_same s, s.center(3)
    refute_same s, result
  end

  def test_odd_remainder_goes_to_right_regardless_of_width_parity
    # 余りが奇数なら右側が 1 文字多い。幅の偶奇に依存しない
    assert_equal "abc ", "abc".center(4)
    assert_equal " ab  ", "ab".center(5)
    assert_equal " a  ", "a".center(4)
    assert_equal "  abc  ", "abc".center(7)
  end

  def test_width_at_or_below_length_returns_same_content
    # 幅が長さ以下（0、負数を含む）なら切り詰めずそのまま
    assert_equal "abc", "abc".center(2)
    assert_equal "abc", "abc".center(0)
    assert_equal "abc", "abc".center(-1)
  end

  def test_multi_char_padstr_repeats_from_each_edge
    # 複数文字の padstr は左右それぞれの端から繰り返し、端数は切り落とす
    assert_equal "_-abc_-_", "abc".center(8, "_-")
    assert_equal "_-_abc_-_", "abc".center(9, "_-")
    assert_equal "abc_-_-_", "abc".ljust(8, "_-")
    assert_equal "_-_-_abc", "abc".rjust(8, "_-")
  end

  def test_empty_string_is_filled_with_padstr
    # 空文字は padstr だけで埋まる
    assert_equal "xxx", "".center(3, "x")
  end

  def test_invalid_arguments
    # 空の padstr は ArgumentError、型が違えば TypeError、Float の幅は切り捨て
    assert_raises(ArgumentError) { "abc".center(8, "") }
    assert_raises(TypeError) { "abc".center(8, 1) }
    assert_raises(TypeError) { "abc".center("8") }
    assert_raises(TypeError) { "abc".center(nil) }
    assert_equal "  abc   ", "abc".center(8.9)
  end

  def test_incompatible_encodings_raise
    # padstr と自身のエンコーディングが互換でなければ Encoding::CompatibilityError。埋め込み不要な幅でも投げる
    sjis = "日".encode("Shift_JIS")
    assert_raises(Encoding::CompatibilityError) { "日本".center(8, sjis) }
    assert_raises(Encoding::CompatibilityError) { "日本".center(1, sjis) }
    assert_equal 8, "abc".center(8, sjis).length
  end

  def test_counts_codepoints
    # 長さはコードポイント単位。全角も 1、結合文字や ZWJ 絵文字列は複数
    assert_equal "**日本**", "日本".center(6, "*")
    assert_equal "*😀*", "😀".center(3, "*")
    assert_equal "*é*", "é".center(4, "*")
    assert_equal 8, "abc".center(8, "日").length
    assert_equal "日日abc日日日", "abc".center(8, "日")
  end

  def test_zero_padding_alternatives
    # Alternatives: 負数のゼロ埋めは rjust だと符号が崩れる。format を使う
    assert_equal "005", "5".rjust(3, "0")
    assert_equal "0-5", "-5".rjust(3, "0")
    assert_equal "-05", format("%03d", -5)
    assert_equal "abc     |", format("%-8s|", "abc")
    assert_equal "     abc|", format("%8s|", "abc")
  end
end
